src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCompilationModuleGroupBase.cs

```diff
@@ -0,0 +1,293 @@
+                    {
+                        try
+                        {
+                            TypeDesc typeFromTypeRef = ecmaModule.GetType(typeRefHandle);
+                            if (!_typeRefsInCompilationModuleSet.ContainsKey(typeFromTypeRef))
+                            {
+                                _typeRefsInCompilationModuleSet.Add(typeFromTypeRef, new ModuleToken(ecmaModule, typeRefHandle));
+                            }
+                        }
+                        catch (TypeSystemException) { }
+                    }
+                }
+            }
+
+            return _typeRefsInCompilationModuleSet.TryGetValue(type, out token);
+        }
+
+        private bool ComputeTypeVersionsWithCode(TypeDesc type)
+        {
+            if (type.IsCanonicalDefinitionType(CanonicalFormKind.Any))
+                return true;
+
+            if (type is MetadataType mdType)
+            {
+                if (!_versionBubbleModuleSet.Contains(mdType.Module))
+                    return false;
+            }
+
+            if (type == type.GetTypeDefinition())
+                return true;
+
+            return ComputeInstantiationVersionsWithCode(type.Instantiation, type);
+        }
+
+        private bool ComputeInstantiationVersionsWithCode(Instantiation inst, TypeSystemEntity entityWithInstantiation)
+        {
+            for (int iInstantiation = 0; iInstantiation < inst.Length; iInstantiation++)
+            {
+                TypeDesc instType = inst[iInstantiation];
+
+                if (!ComputeInstantiationTypeVersionsWithCode(this, instType))
+                {
+                    if (instType.IsPrimitive)
+                    {
+                        // Primitive type instantiations are only instantiated in the module of the generic defining type
+                        // if the generic does not apply interface constraints to that type parameter, or if System.Private.CoreLib is part of the version bubble
+
+                        Instantiation entityDefinitionInstantiation;
+                        if (entityWithInstantiation is TypeDesc type)
+                        {
+                            entityDefinitionInstantiation = type.GetTypeDefinition().Instantiation;
+                        }
+                        else
+                        {
+                            entityDefinitionInstantiation = ((MethodDesc)entityWithInstantiation).GetTypicalMethodDefinition().Instantiation;
+                        }
+
+                        GenericParameterDesc genericParam = (GenericParameterDesc)entityDefinitionInstantiation[iInstantiation];
+                        if (genericParam.HasReferenceTypeConstraint)
+                            return false;
+
+                        // This checks to see if the type constraints list is empty
+                        if (genericParam.TypeConstraints.GetEnumerator().MoveNext())
```

Comment:
Does this enumerate a constraint like IEquatable<int>? If so, shouldn't we detect if SPC is in the version bubble before returning false (also what the comment at line 244 says)

Reply:
We did that already. That is what line 239+241 did. We only hit this case when compiling without corelib in the bubble.